File: Top 100 Codes/06-Advanced Codes related to Arrays/10-Kadane’s Algorithm/KadanesAlgorithm.py
def kadane_algorithm(nums):
    max_so_far = float('-inf')
    max_ending_here = 0

    for i in range(len(nums)):
        max_ending_here = max(nums[i], max_ending_here + nums[i])
        max_so_far = max(max_so_far, max_ending_here)

    return max_so_far


nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
print("Maximum subarray sum is", kadane_algorithm(nums))

File: Top 100 Codes/06-Advanced Codes related to Arrays/10-Kadane’s Algorithm/KadanesAlgorithm.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

int kadaneAlgorithm(const vector<int>& nums) {
    int maxSoFar = INT_MIN;
    int maxEndingHere = 0;

    for (int i = 0; i < nums.size(); i++) {
        maxEndingHere = max(nums[i], maxEndingHere + nums[i]);
        maxSoFar = max(maxSoFar, maxEndingHere);
    }

    return maxSoFar;
}

int main() {
    vector<int> nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    cout << "Maximum subarray sum is " << kadaneAlgorithm(nums) << endl;
    return 0;
}
